cpp/src/arrow/compute/dictionary_decode.cc

```diff
@@ -0,0 +1,95 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/dictionary_decode.h"
+
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "arrow/compute/cast.h"
+#include "arrow/compute/cast_internal.h"
+#include "arrow/compute/exec.h"
+#include "arrow/compute/function_internal.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/codegen_internal.h"
+#include "arrow/compute/registry.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/reflection_internal.h"
+
+namespace arrow {
+
+namespace compute {
+namespace internal {
+
+namespace {
+
+const FunctionDoc dictionary_decode_doc{"Decodes a DictionaryArray to an Array",
+                                        "The Function will call cast to really decode.",
+                                        {"dictionary_array"}};
+class DictionaryDecodeMetaFunction : public MetaFunction {
+ public:
+  DictionaryDecodeMetaFunction()
+      : MetaFunction("dictionary_decode", Arity::Unary(), dictionary_decode_doc) {}
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    if (args[0].type() == nullptr || args[0].type()->id() != Type::DICTIONARY) {
+      return Status::TypeError("Expected a DictonaryArray");
+    }
+
+    if (args[0].is_array()) {
+      ARROW_CHECK_NE(args[0].array()->dictionary, nullptr);
```

Comment:
Instead of grabbing the dictionary from the array/chunk it would probably be easier to grab it from the datum's type property.  Then you don't even need the if/else loop I think:

```
  DictionaryType* dict_type = checked_cast<DictionaryType>(args[0].type().get());
  CastOptions cast_options = CastOptions::Safe(dict_type->value_type());
```